lib/container/HoistContainerReferencesPlugin.js

```diff
@@ -0,0 +1,205 @@
+
+				for (const module of allReferencedModules) {
+					if (!chunkGraph.isModuleInChunk(module, runtimeChunk)) {
+						chunkGraph.connectChunkAndModule(runtimeChunk, module);
+					}
+				}
+			}
+			this.cleanUpChunks(compilation, allReferencedModules);
+		}
+	}
+
+	/**
+	 * Clean up chunks by disconnecting unused modules.
+	 * @param {import("webpack").Compilation} compilation The webpack compilation instance.
+	 * @param {Set<import("webpack").Module>} modules Set of modules to clean up.
+	 */
+	cleanUpChunks(compilation, modules) {
+		const { chunkGraph } = compilation;
+		for (const module of modules) {
+			for (const chunk of chunkGraph.getModuleChunks(module)) {
+				if (!chunk.hasRuntime()) {
+					chunkGraph.disconnectChunkAndModule(chunk, module);
+					if (
+						chunkGraph.getNumberOfChunkModules(chunk) === 0 &&
+						chunkGraph.getNumberOfEntryModules(chunk) === 0
+					) {
+						chunkGraph.disconnectChunk(chunk);
+						compilation.chunks.delete(chunk);
+						if (chunk.name) {
+							compilation.namedChunks.delete(chunk.name);
+						}
+					}
+				}
+			}
+		}
+		modules.clear();
+	}
+
+	/**
+	 * Helper method to get runtime chunks from the compilation.
+	 * @param {import("webpack").Compilation} compilation The webpack compilation instance.
+	 * @returns {Set<import("webpack").Chunk>} Set of runtime chunks.
+	 */
+	getRuntimeChunks(compilation) {
+		const runtimeChunks = new Set();
+		const entries = compilation.entrypoints;
+
+		for (const entrypoint of entries.values()) {
+			const runtimeChunk = entrypoint.getRuntimeChunk();
+			if (runtimeChunk) {
+				runtimeChunks.add(runtimeChunk);
+			}
+		}
+		return runtimeChunks;
+	}
+}
+
+/**
+ * Helper method to collect all referenced modules recursively.
+ * @param {import("webpack").Compilation} compilation The webpack compilation instance.
+ * @param {import("webpack").Module} module The module to start collecting from.
+ * @param {string} type The type of modules to collect ("initial", "external", or "all").
+ * @returns {Set<import("webpack").Module>} Set of collected modules.
+ */
+function getAllReferencedModules(compilation, module, type) {
+	const collectedModules = new Set([module]);
+	const visitedModules = new WeakSet([module]);
+	const stack = [module];
+
+	while (stack.length > 0) {
+		const currentModule = stack.pop();
+		if (!currentModule) continue;
+
+		const mgm = compilation.moduleGraph._getModuleGraphModule(currentModule);
+		if (mgm && mgm.outgoingConnections) {
+			for (const connection of mgm.outgoingConnections) {
+				const connectedModule = connection.module;
+
+				// Skip if module has already been visited
+				if (!connectedModule || visitedModules.has(connectedModule)) {
+					continue;
+				}
+
+				// Handle 'initial' type (skipping async blocks)
+				if (type === "initial") {
+					const parentBlock = compilation.moduleGraph.getParentBlock(
+						connection.dependency
+					);
+					if (parentBlock instanceof AsyncDependenciesBlock) {
```

Comment:
skip async imports since chunk handler will be able to get those dependencies ahead of time,  not eager